ffi/passmanagers.cpp

```diff
@@ -24,164 +41,380 @@ using namespace llvm;
 extern "C" {
 
 API_EXPORT(void)
-LLVMPY_SetTimePasses(bool enable){
-  TimePassesIsEnabled = enable;
-}
+LLVMPY_SetTimePasses(bool enable) { TimePassesIsEnabled = enable; }
 
 API_EXPORT(void)
 LLVMPY_ReportAndResetTimings(const char **outmsg) {
-    std::string osbuf;
-    raw_string_ostream os(osbuf);
-    reportAndResetTimings(&os);
-    os.flush();
-    *outmsg = LLVMPY_CreateString(os.str().c_str());
+  std::string osbuf;
+  raw_string_ostream os(osbuf);
+  reportAndResetTimings(&os);
+  os.flush();
+  *outmsg = LLVMPY_CreateString(os.str().c_str());
 }
 
-
 API_EXPORT(LLVMPassManagerRef)
-LLVMPY_CreatePassManager()
-{
-    return LLVMCreatePassManager();
-}
+LLVMPY_CreatePassManager() { return LLVMCreatePassManager(); }
 
 API_EXPORT(void)
-LLVMPY_DisposePassManager(LLVMPassManagerRef PM)
-{
-    return LLVMDisposePassManager(PM);
+LLVMPY_DisposePassManager(LLVMPassManagerRef PM) {
+  return LLVMDisposePassManager(PM);
 }
 
 API_EXPORT(LLVMPassManagerRef)
-LLVMPY_CreateFunctionPassManager(LLVMModuleRef M)
-{
-    return LLVMCreateFunctionPassManagerForModule(M);
+LLVMPY_CreateFunctionPassManager(LLVMModuleRef M) {
+  return LLVMCreateFunctionPassManagerForModule(M);
 }
 
 API_EXPORT(int)
-LLVMPY_RunPassManager(LLVMPassManagerRef PM,
-                      LLVMModuleRef M)
-{
-    return LLVMRunPassManager(PM, M);
+LLVMPY_RunPassManager(LLVMPassManagerRef PM, LLVMModuleRef M) {
+  return LLVMRunPassManager(PM, M);
 }
 
 API_EXPORT(int)
-LLVMPY_RunFunctionPassManager(LLVMPassManagerRef PM,
-                              LLVMValueRef F)
-{
-    return LLVMRunFunctionPassManager(PM, F);
+LLVMPY_RunFunctionPassManager(LLVMPassManagerRef PM, LLVMValueRef F) {
+  return LLVMRunFunctionPassManager(PM, F);
 }
 
 API_EXPORT(int)
-LLVMPY_InitializeFunctionPassManager(LLVMPassManagerRef FPM)
-{
-    return LLVMInitializeFunctionPassManager(FPM);
+LLVMPY_InitializeFunctionPassManager(LLVMPassManagerRef FPM) {
+  return LLVMInitializeFunctionPassManager(FPM);
 }
 
 API_EXPORT(int)
-LLVMPY_FinalizeFunctionPassManager(LLVMPassManagerRef FPM)
-{
-    return LLVMFinalizeFunctionPassManager(FPM);
+LLVMPY_FinalizeFunctionPassManager(LLVMPassManagerRef FPM) {
+  return LLVMFinalizeFunctionPassManager(FPM);
+}
+
+API_EXPORT(void)
+LLVMPY_AddAAEvalPass(LLVMPassManagerRef PM) {
+  unwrap(PM)->add(llvm::createAAEvalPass());
+}
+
+API_EXPORT(void)
+LLVMPY_AddBasicAAPass(LLVMPassManagerRef PM) {
+  unwrap(PM)->add(llvm::createBasicAAWrapperPass());
+}
+
+API_EXPORT(void)
+LLVMPY_AddDependenceAnalysisPass(LLVMPassManagerRef PM) {
+  unwrap(PM)->add(llvm::createDependenceAnalysisWrapperPass());
+}
+
+API_EXPORT(void)
+LLVMPY_AddCallGraphViewerPass(LLVMPassManagerRef PM) {
+  unwrap(PM)->add(llvm::createCallGraphViewerPass());
+}
+
+API_EXPORT(void)
+LLVMPY_AddDotCallGraphPass(LLVMPassManagerRef PM) {
+  unwrap(PM)->add(llvm::createCallGraphDOTPrinterPass());
```

Comment:
At some point it might be good to wire up the output "files" as python strings.

Reply:
What would the lifecycle management be on that? Most of the interfaces require `ostream&`, so we would need somewhere to store that data, possibly for the lifetime of the pass manager.